sound/soc/sof/intel/hda.c

```diff
@@ -519,6 +519,15 @@ static int hda_init(struct snd_sof_dev *sdev)
 	/* HDA base */
 	sdev->bar[HDA_DSP_HDA_BAR] = bus->remap_addr;
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	/* init i915 and HDMI codecs */
+	ret = hda_codec_i915_init(sdev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: init i915 and HDMI codec failed\n");
+		return ret;
+	}
+#endif
+
 	/* get controller capabilities */
 	ret = hda_dsp_ctrl_get_caps(sdev);
 	if (ret < 0)
```

Comment:
@ranj063 @plbossart  I'm wondering why we are doing hda_dsp_ctrl_get_caps() in hda_init()..? This is fairly confusing as there is a separate hda_init_caps(). I was thinking should we keep hda_init() as a simple function that just allocates the kernel structures but does not do any hw i/o and then have a separate function that does hda_dsp_ctrl_get_caps() and friends. So far so good, but then I hit the naming issue as we already have hda-init_caps() which does much more.

So something like:
 - hda_init()  -> remove all hw i/o
 - hda_init_caps() -> do the i915 init and hda_dsp_ctrl_get_caps() 
 - hda_init_probe() -> the current hda_init_caps() which initializes bus'es and runs probe for HDA and SDW

...?

Reply:
wow.  

> @plbossart wrote:
> 
> > the link resets in get_caps (" ASoC: SOF: Intel: hda-ctrl: add reset cycle before parsing capabilities") > were added to work around SoundWire-related issues, we never root-caused why they were
> > needed. if they cause problems for HDMI, you can consider removing this...
> 
> Ack. We actually tried but still got the error. Based on test results so far, link needs to be functional when get_caps is done (i.e. the i915 get_power needs to be done before, which is a bit surprising, but seems to be so). Could also explain your issue (that you had to do a link reset before get_caps). Plus azx is doing it this way.

I have no idea what the issue is. The point is that without the reset cycle, we could not read the capabilities on customer SoundWire-based devices, so we failed to probe. We never had such issues on Intel RVPs. I suspect a BIOS dependency or something.

Reply:
I never realized there were two calls to that get_caps function. That's not right...